### PR TITLE
[DQL2-13] CKAN Harvest - Install and configure

### PR DESCRIPTION
**Pull request by salsa-nathan**
- Added `supervisor-ckan-harvest-gather.conf` for harvest gather consumer background process
- Added `supervisor-ckan-harvest-fetch.conf` for harvest fetch consumer background process
- Copy harvest supervisor conf files in `/recipes/ckanweb-deploy.rb`
- Added ckanext-harvest `initdb` paster command to `recipes/ckanweb-deploy-exts.rb`
- Added ckanext-harvest `harvester run` cron job to `recipes/ckanweb-deploy.rb`

cc: @duttonw & @ThrawnCA

### CODE REVIEW

File: files/default/supervisor-ckan-harvest-gather.conf [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+
+; Start just a single gather consumer.
+numprocs=1
+process_name=%(program_name)s-%(process_num)02d
+
+
+; Log files.
+stdout_logfile=/var/log/ckan/ckan-harvest-gather.log
+stderr_logfile=/var/log/ckan/ckan-harvest-gather.log
```
Comment (duttonw) [inline files/default/supervisor-ckan-harvest-gather.conf:32]: add note that these needs to by copied to s3 and setup to cloudwatch logs inside opsworks

File: files/default/supervisor-ckan-harvest-fetch.conf [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+user=ckan
+
+
+; Start just a single fetch consumer.
+numprocs=1
+process_name=%(program_name)s-%(process_num)02d
+
+
+; Log files.
+stdout_logfile=/var/log/ckan/ckan-harvest-fetch.log
```
Comment (duttonw) [inline files/default/supervisor-ckan-harvest-fetch.conf:31]: add note that these needs to by copied to s3 and setup to cloudwatch logs inside opsworks

File: recipes/ckanweb-deploy.rb [hunk 3]
```diff
@@ -275,6 +275,20 @@
 	mode "0744"
 end
 
+cookbook_file "/etc/supervisor/conf.d/supervisor-ckan-harvest-gather.conf" do
```
Comment (duttonw) [inline recipes/ckanweb-deploy.rb:278]: is this ok to run on N+1 nodes? same for fetch.

File: files/default/supervisor-ckan-harvest-gather.conf (hunk 1 above)
Comment (ThrawnCA) [reply files/default/supervisor-ckan-harvest-gather.conf:32]: They should be picked up by the existing processes. See https://github.com/qld-gov-au/opswx-ckan-cookbook/blob/master/files/default/ckan-logrotate which rotates `/var/log/ckan/*log`

File: files/default/supervisor-ckan-harvest-gather.conf [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+; ====================================================================
+; Supervisor configuration for CKAN Harvest background gather consumer
+; ====================================================================
+
+; 1. Copy this file to /etc/supervisor/conf.d
+; 2. Make sure the paths below match your setup
+
+
+[program:ckan-harvest-gather]
```
Comment (ThrawnCA) [inline files/default/supervisor-ckan-harvest-gather.conf:15]: This should probably be combined with the other Supervisor config file.

File: recipes/ckanweb-deploy-exts.rb [hunk 5]
```diff
@@ -250,6 +250,12 @@
 			only_if { "#{pluginname}".eql? 'ytp-comments' }
 		end
 
+		execute "Harvest CKAN ext database init" do
+			user "#{account_name}"
+			command "#{virtualenv_dir}/bin/paster --plugin=ckanext-harvest harvester initdb -c #{config_dir}/production.ini || echo 'Ignoring expected error, see https://github.com/frictionlessdata/ckanext-validation/issues/44'"
```
Comment (ThrawnCA) [inline recipes/ckanweb-deploy-exts.rb:255]: Comment refers to the wrong extension. Also, does this one fail gracefully in the same way?

File: recipes/ckanweb-deploy.rb [hunk 6]
```diff
@@ -322,3 +336,10 @@
 	group "root"
 	mode "0755"
 end
+
+file "/etc/cron.hourly/ckan-harvest-run" do
+	content "/usr/local/bin/pick-job-server.sh && echo '{}' | #{paster} --plugin=ckanext-harvest harvester run -c #{config_file} 2>&1 > /dev/null\n"
```
Comment (ThrawnCA) [inline recipes/ckanweb-deploy.rb:341]: Do we need to use `pick-job-server.sh` here, or is it ok for each server to run it?

**Comment by ThrawnCA** [comment]
Where is the configuration of the Redis URL (`ckan.harvest.mq.hostname` etc)? We need to point it at Elasticache.

File: recipes/ckanweb-deploy.rb (hunk 6 above)
Comment (salsa-nathan) [reply recipes/ckanweb-deploy.rb:341]: I added `pick-job-server.sh` based on your comment in the DQL2-13 story: https://salsadigital.atlassian.net/browse/DQL2-13?focusedCommentId=134766 -- should I remove it?

**Comment by salsa-nathan** [comment]
> Where is the configuration of the Redis URL (`ckan.harvest.mq.hostname` etc)? We need to point it at Elasticache.

Please see my note in DQL2-13: https://salsadigital.atlassian.net/browse/DQL2-13?focusedCommentId=135134

"NOTE: I didn't add anything harvest specific to /templates/default/ckan_properties.ini.erb as it looks like the harvest configuration settings already exist, e.g.

https://github.com/qld-gov-au/opswx-ckan-cookbook/blob/develop/templates/default/ckan_properties.ini.erb#L125
"

File: recipes/ckanweb-deploy-exts.rb (hunk 5 above)
Comment (salsa-nathan) [reply recipes/ckanweb-deploy-exts.rb]: sorry, copied from the example above - in which case I've removed those as well as they would be incorrect

File: recipes/ckanweb-deploy.rb (hunk 3 above)
Comment (salsa-nathan) [reply recipes/ckanweb-deploy.rb:278]: It shouldn't need to run on N+1 nodes - is there a way to restrict it to a single node?

Comment (ThrawnCA) [reply recipes/ckanweb-deploy.rb:278]: If we don't run it on every server, then we can't freely fail over from one server to another. I would like to run it everywhere if possible.

File: recipes/ckanweb-deploy.rb (hunk 6 above)
Comment (ThrawnCA) [reply recipes/ckanweb-deploy.rb:341]: If the job can run simultaneously on multiple servers without conflict and without doubling up, then I'd like to let it do that. Picking a single server is really only for cases where race conditions are a problem.

File: recipes/ckanweb-deploy-exts.rb [hunk 7]
```diff
@@ -243,13 +243,19 @@
...
+				#{virtualenv_dir}/bin/paster --plugin=ckanext-ytp-comments initdb -c #{config_dir}/production.ini || echo 'Ignoring expected error'
+				#{virtualenv_dir}/bin/paster --plugin=ckanext-ytp-comments init_notifications_db -c #{config_dir}/production.ini || echo 'Ignoring expected error'
+				#{virtualenv_dir}/bin/paster --plugin=ckanext-ytp-comments updatedb -c #{config_dir}/production.ini || echo 'Ignoring expected error'
 			EOS
 			only_if { "#{pluginname}".eql? 'ytp-comments' }
 		end
 
+		execute "Harvest CKAN ext database init" do
+			user "#{account_name}"
+			command "#{virtualenv_dir}/bin/paster --plugin=ckanext-harvest harvester initdb -c #{config_dir}/production.ini || echo 'Ignoring expected error"
```
Comment (ThrawnCA) [inline recipes/ckanweb-deploy-exts.rb:255]: Missing closing single quote after warning message.

**Pull request merged**